### PR TITLE
[SPARK-37728][SQL] Reading nested columns with ORC vectorized reader can cause ArrayIndexOutOfBoundsException

### PR DESCRIPTION
**Pull request by yimin-yang**
### What changes were proposed in this pull request?

When an OrcColumnarBatchReader is created, method initBatch will be called only once. In method initBatch:

`orcVectorWrappers[i] = OrcColumnVectorUtils.toOrcColumnVector(dt, wrap.batch().cols[colId]);`

When the second argument of toOrcColumnVector is a ListColumnVector/MapColumnVector, orcVectorWrappers[i] is initialized with the ListColumnVector or MapColumnVector's offsets and lengths. 

However, when method nextBatch of OrcColumnarBatchReader is called, method ensureSize of ColumnVector (and its subclasses, like MultiValuedColumnVector) could be called, then the ListColumnVector/MapColumnVector's offsets and lengths could refer to new array objects. This could result in the ArrayIndexOutOfBoundsException. 

This PR makes OrcArrayColumnVector.getArray and OrcMapColumnVector.getMap always get offsets and lengths from the underlying ColumnVector, which can resolve this issue.

### Why are the changes needed?

Bugfix

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Pass the CIs with the newly added test case.

### CODE REVIEW

**Comment by yym1995** [comment]
@c21  Could you take a look when you are free? Thanks! Looking forward to your feedback.

**Comment by c21** [comment]
@yym1995 thank you submitting a fix! Could you help add a unit test case as well?

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
cc @dongjoon-hyun FYI

**Comment by yym1995** [comment]
> @yym1995 thank you submitting a fix! Could you help add a unit test case as well?

I just added a unit test case. Please take a look, thanks!

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcArrayColumnVector.java [hunk 1]
```diff
@@ -31,26 +32,27 @@
  */
 public class OrcArrayColumnVector extends OrcColumnVector {
   private final OrcColumnVector data;
-  private final long[] offsets;
-  private final long[] lengths;
+  private ListColumnVector listData;
 
   OrcArrayColumnVector(
```
Comment (c21) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcArrayColumnVector.java:37]: I think we don't need to store another copy of `vector` here. We can change [`OrcColumnVector.baseData`](https://github.com/apache/spark/blob/master/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnVector.java#L32) from `private` to `protected`, and just use `baseData` to get offset and length. So how about defining the constructor like below:

```java
OrcArrayColumnVector(
    DataType type,
    ListColumnVector vector,
    OrcColumnVector data) {

  super(type, vector);

  this.data = data;
}
```

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcMapColumnVector.java [hunk 2]
```diff
@@ -32,28 +33,30 @@
 public class OrcMapColumnVector extends OrcColumnVector {
   private final OrcColumnVector keys;
   private final OrcColumnVector values;
-  private final long[] offsets;
-  private final long[] lengths;
+  private MapColumnVector mapData;
 
   OrcMapColumnVector(
```
Comment (c21) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcMapColumnVector.java:38]: Similarly we can define `OrcMapColumnVector` constructor as below:

```java
OrcMapColumnVector(
    DataType type,
    MapColumnVector vector,
    OrcColumnVector keys,
    OrcColumnVector values) {

  super(type, vector);

  this.keys = keys;
  this.values = values;
}
```

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcMapColumnVector.java [hunk 3]
```diff
@@ -32,28 +33,30 @@
...
+      mapData = (MapColumnVector) vector;
+    } else {
+      throw new UnsupportedOperationException();
+    }
   }
 
   @Override
   public ColumnarMap getMap(int ordinal) {
-    return new ColumnarMap(keys, values, (int) offsets[ordinal], (int) lengths[ordinal]);
+    return new ColumnarMap(keys, values, (int) mapData.offsets[ordinal],
```
Comment (c21) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcMapColumnVector.java:58]:
```java
public ColumnarMap getMap(int ordinal) {
  int offset = (int) ((MapColumnVector) baseData).offsets[ordinal];
  int length = (int) ((MapColumnVector) baseData).lengths[ordinal];
  return new ColumnarMap(keys, values, offset, length);
}
```

**Review by c21 — COMMENTED** [review]
Thank you @yym1995 for the fix! Having some comments for code structure, but the fix is valid.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcArrayColumnVector.java [hunk 4]
```diff
@@ -31,26 +32,27 @@
...
+      listData = (ListColumnVector) vector;
+    } else {
+      throw new UnsupportedOperationException();
+    }
   }
 
   @Override
   public ColumnarArray getArray(int rowId) {
-    return new ColumnarArray(data, (int) offsets[rowId], (int) lengths[rowId]);
+    return new ColumnarArray(data, (int) listData.offsets[rowId], (int) listData.lengths[rowId]);
```
Comment (c21) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcArrayColumnVector.java:55]: We can just use `baseData` here:

```java
public ColumnarArray getArray(int rowId) {
  int offset = (int) ((ListColumnVector) baseData).offsets[rowId];
  int length = (int) ((ListColumnVector) baseData).lengths[rowId];
  return new ColumnarArray(data, offset, length);
}
```

**Comment by yym1995** [comment]
@c21 Thank you for the feedback! I have already changed the code structure.

**Comment by yym1995** [comment]
@dongjoon-hyun This PR fixed a bug in ORC vectorized reader. @c21 has reviewed this PR, and I have improved the code according to the feedback. I was wondering if you could merge this PR, thanks!

**Comment by c21** [comment]
LGTM with pending CI tests. cc @viirya and @cloud-fan as well.

**Comment by c21** [comment]
retest this please

**Comment by cloud-fan** [comment]
The fix LGTM. Please click the failed Github Action, follow the instructions to fix the issues.

**Review by viirya — APPROVED** [review]
The fix looks good. Pending CI.

**Comment by viirya** [comment]
GA seems unstable. You can submit an empty commit to re-trigger it.

**Comment by viirya** [comment]
> Tested manually. Just follow the bug reproduction steps in https://issues.apache.org/jira/browse/SPARK-37728 .

As you added an unit test, you can modify the PR description too.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CI)

**Comment by yym1995** [comment]
Now all checks have passed. cc @cloud-fan @dongjoon-hyun @HyukjinKwon @viirya

**Comment by dongjoon-hyun** [comment]
I revised the PR description. Merged to master.
Thank you, @yym1995 and all!

**Comment by dongjoon-hyun** [comment]
Welcome to the Apache Spark community, @yym1995 . 
I added you to the Apache Spark contributor group and assigned SPARK-37728 to you.
Could you make backporting PRs to `branch-3.2` and `branch-3.1`? We need to pass the UTs on those branches.

**Comment by yym1995** [comment]
> Welcome to the Apache Spark community, @yym1995 . I added you to the Apache Spark contributor group and assigned [SPARK-37728](https://issues.apache.org/jira/browse/SPARK-37728) to you. Could you make three backporting PRs to `branch-3.2` and `branch-3.1` and `branch-3.0`? We need to pass the UTs on those branches.

OK, will do.

**Comment by c21** [comment]
@dongjoon-hyun - thanks for merging. We only need backport to branch-3.2, right? The related code was only introduced in 3.2 branch (in PR for ORC vectorized reader of nested column)

**Comment by dongjoon-hyun** [comment]
Yes, we only need to backport to the branch which the test case fails.

BTW, @yym1995 , if `branch-3.2` is the only branch, you should remove `Affected Version` of SPARK-37728. Currently, it has 3.0.2 and 3.1.2.

**Pull request closed without merging**